users/views.py

```diff
@@ -1,3 +1,103 @@
+    def validate_user_input(self, data):
+        self.validate_email(data['email'])
+        self.validate_password(data['password'])
+        
+    def validate_email(self, email):
+        if not re.match('^[a-zA-Z0-9+-_.]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$', email):
+            
+            return JsonResponse({'message': 'EMAIL_NOT_VALID'}, status=400)
+
+    def validate_password(self, password):
+        if not re.match('^(?=.*[A-Za-z])(?=.*\d)(?=.*[?!@#$%*&])[A-Za-z\d?!@#$%*&]{8,}$', password):
+            
+            return JsonResponse({'message': 'PASSWORD_NOT_VALID'}, status=400)
+
+    def hash_pw(self, password):
+        return bcrypt.hashpw(password.encode('utf-8'), bcrypt.gensalt()).decode('utf-8')
+
+    def post(self, request):
+        try:
+            data     = json.loads(request.body)
+            name     = data['name']
+            email    = data['email']
+            password = data['password']
+            
+            self.validate_user_input(data)
+            
+            hashed_password = self.hash_pw(password)
+            
+            User.objects.create(name=name, password=hashed_password, email=email)
+            
+            return JsonResponse({'message': 'SUCCESS'}, status=201)
+
+        except KeyError:
+            return JsonResponse({'message': 'KEY_ERROR'}, status=400)
+
+class SignInView(View):
+    def post(self, request):
+        try:
+            data         = json.loads(request.body)
+            email        = data['email']
+            password     = data['password']
+            user         = User.objects.get(email=email)
+            access_token = jwt.encode({'user_id': user.id, 'exp': datetime.utcnow() + timedelta(days=7)}, SECRET_KEY, ALGORITHM)
+
+            if user.deleted_at != None:
+                return JsonResponse({'message': 'UNACTIVATED_USER'}, status=403)
+
+            if not bcrypt.checkpw(password.encode('utf-8'), user.password.encode('utf-8')):
+                return JsonResponse({'message': 'INVALID_PASSWORD'}, status=404)
+
+            return JsonResponse({'message': 'SUCCESS', 'access_token': access_token}, status=200)
+
+        except jwt.ExpiredSignatureError:
+            return JsonResponse({'message': 'EXPIRED_TOKEN'}, status = 400)
+
+        except KeyError:
+            return JsonResponse({'message': 'KEY_ERROR'}, status=400)
+
+        except User.DoesNotExist:
+            return JsonResponse({'message': 'INVALID_USER'}, status=404)
+        
+class KakaoLoginView(View):
+    def get(self, request):
+        try:
+            kakao_token = request.headers.get('Authorization', None)
+            
+            if kakao_token == None:
+                return JsonResponse({'message' : 'INVALID_TOKEN'})
+            
+            kakao_url = 'https://kapi.kakao.com/v2/user/me'
+            headers   = {'Authorization': f'Bearer {kakao_token}'}
+            
+            kakao_response = requests.get(kakao_url, headers = headers).json()
+            
```

Comment:
1. requests 모듈을 사용하실 때에는 timeout 옵션을 반드시 설정해주셔야 합니다.
그 이유는 timeout option을 추가하지 않으면 서버의 응답을 계속해서 기다리고 있는 문제가 발생 합니다. timeout 옵션을 추가해 주세요.
```suggestion
            kakao_response = requests.get(kakao_url, headers = headers, timeout=5).json()
            
```

2.  kakao로부터 항상 올바른 응답(200)을 받는다는 보장은 없습니다. 카카오로부터 에러코드를 받을 경우에 대해서 어떻게 처리해야 할지 생각해보고 코드에 예외처리 추가해 주세요.

Reply:
리뷰 반영하여 timeout, key_error 추가하였습니다.